db/db_impl/db_impl.cc

```diff
@@ -1532,20 +1532,26 @@ void DBImpl::MarkLogsSynced(uint64_t up_to, bool synced_dir,
     auto& wal = *it;
     assert(wal.IsSyncing());
 
-    if (logs_.size() > 1) {
+    if (wal.number < logs_.back().number) {
+      // Inactive WAL
       if (immutable_db_options_.track_and_verify_wals_in_manifest &&
           wal.GetPreSyncSize() > 0) {
         synced_wals->AddWal(wal.number, WalMetadata(wal.GetPreSyncSize()));
       }
-      logs_to_free_.push_back(wal.ReleaseWriter());
-      it = logs_.erase(it);
+      if (wal.GetPreSyncSize() == wal.writer->file()->GetFlushedSize()) {
```

Comment:
Done, though slightly differently as that gives me fear of infinite loop

Reply:
@ajkr It might be noob question. How is it ensured that no other thread is writing to this wal file at this point.

Reply:
It looks like `wal.number < logs_.back().number` tells us the log is inactive because a newer one exists, and we only append to the newest (active) WAL.